Documentation/teaching/lectures/smp.rst

```diff
@@ -0,0 +1,1156 @@
+                          lock        lock      lock
+
+
+
+Conceptually, when a new CPU core tries to acquire the lock and it
+fails it will add its private lock to the list of waiting CPU
+cores. When the lock owner exits the critical section it unlocks the
+next lock in the list, if any.
+
+While a read spin optimized spin lock reduces most of the cache
+invalidation operations, the lock owner can still generate cache
+invalidate operations due to writes to data structures close to the
+lock. This in turn generates memory traffic on subsequent reads on the
+spinning cores.
+
+Hence, queued spin locks scale much better for large number of cores
+as is the case for NUMA systems. And since they have similar fairness
+properties as the ticket lock it is the preferred implementation on
+the x86 architecture.
+
+
+Process and Interrupt Context Synchronization
+=============================================
+
+Accessing shared data from both process and interrupt context is a
+relatively common scenario. On single core systems we can do this by
+disabling interrupts, but that won't work on multi-core systems,
+as we can have the process running on one CPU core and the interrupt
+context running on a different CPU core.
+
+Using a spin lock, which was designed for multi-processor systems,
+seems like the right solution, but doing so will can cause common
+deadlock condition, as detailed by the following scenario:
+
+
+.. slide:: Process and Interrupt Handler Synchronization Deadlock
+   :inline-contents: True
+   :level: 2
+
+   * In the process context we take the spin lock
+
+   * An interrupt occurs and it is scheduled on the same CPU core
+
+   * The interrupt handler runs and tries to take the spin lock
+
+   * The current CPU will deadlock
+
+
+To avoid this issue a two fold approach is used:
+
+
+.. slide:: Interrupt Synchronization for SMP
+   :inline-contents: True
+   :level: 2
+
+   * In process context: disable interrupts and acquire a spin lock;
+     this will protect both against interrupt or other CPU cores race
+     conditions (:c:func:`spin_lock_irqsave` and
+     :c:func:`spin_lock_restore` combine the two operations)
+
+   * In interrupt context: take a spin lock; this will will protect
+     against race conditions with other interrupt handlers or process
+     context running on different processors
+
+
+We have the same issue for other interrupt context handlers such as
+softirqs, tasklets or timers and while disabling interrupts might
+work, it is recommended to use dedicated APIs:
+
+.. slide:: Bottom-Half Synchronization for SMP
+   :inline-contents: True
+   :level: 2
+
+   * In process context use :c:func:`spin_lock_bh` (which combines
+   :c:func:`local_bh_disable` and :c:func:`spin_lock`) and
+   :c:func:`spin_unlock_bh` (which combines :c:func:`spin_unlock` and
+   :c:func:`local_bh_enable`)
+
+
+   * In bottom half context use: :c:func:`spin_lock` and
+   :c:func:`spin_unlock` (or :c:func:`spin_lock_irqsave` and
```

Comment:
This appears as a new line in the HTML output.

Reply:
This looks like a sphinx bug, I tried a few workarounds but I could not fixed it :(